onnx/test/test_backend_reference.py

```diff
@@ -0,0 +1,152 @@
+
+    def run(self, inputs, **kwargs):
+        if isinstance(inputs, list):
+            feeds = {}
+            for i, inp in enumerate(self._session.input_names):
+                if i >= len(inputs):
+                    break
+                feeds[inp] = inputs[i]
+        elif isinstance(inputs, dict):
+            feeds = inputs
+        elif isinstance(inputs, numpy.ndarray):
+            names = self._session.input_names
+            if len(names) != 1:
+                raise RuntimeError(f"Expecting one input not {len(names)}.")
+            feeds = {names[0]: inputs}
+        else:
+            raise TypeError(f"Unexpected input type {type(inputs)!r}.")
+        outs = self._session.run(None, feeds)
+        return outs
+
+
+class ReferenceEvaluatorBackend(onnx.backend.base.Backend):
+    @classmethod
+    def is_opset_supported(cls, model):  # pylint: disable=unused-argument
+        return True, ""
+
+    @classmethod
+    def supports_device(cls, device: str) -> bool:
+        d = Device(device)
+        return d.type == DeviceType.CPU  # type: ignore[no-any-return]
+
+    @classmethod
+    def create_inference_session(cls, model):
+        return ReferenceEvaluator(model)
+
+    @classmethod
+    def prepare(
+        cls, model: Any, device: str = "CPU", **kwargs: Any
+    ) -> ReferenceEvaluatorBackendRep:
+        # if isinstance(model, ReferenceEvaluatorBackendRep):
+        #    return model
+        if isinstance(model, ReferenceEvaluator):
+            return ReferenceEvaluatorBackendRep(model)
+        if isinstance(model, (str, bytes, ModelProto)):
+            inf = cls.create_inference_session(model)
+            return cls.prepare(inf, device, **kwargs)
+        raise TypeError(f"Unexpected type {type(model)} for model.")
+
+    @classmethod
+    def run_model(cls, model, inputs, device=None, **kwargs):
+        rep = cls.prepare(model, device, **kwargs)
+        return rep.run(inputs, **kwargs)
+
+    @classmethod
+    def run_node(cls, node, inputs, device=None, outputs_info=None, **kwargs):
+        raise NotImplementedError("Unable to run the model node by node.")
+
+
+backend_test = onnx.backend.test.BackendTest(ReferenceEvaluatorBackend, __name__)
+
+if os.getenv("APPVEYOR"):
+    backend_test.exclude("(test_vgg19|test_zfnet)")
+if platform.architecture()[0] == "32bit":
+    backend_test.exclude("(test_vgg19|test_zfnet|test_bvlc_alexnet)")
+if platform.system() == "Windows":
+    backend_test.exclude("test_sequence_model")
+
+# The following tests are not supported.
+backend_test.exclude(
+    "(test_gradient"
+    "|test_if_opt"
+    "|test_loop16_seq_none"
+    "|test_range_float_type_positive_delta_expanded"
+    "|test_range_int32_type_negative_delta_expanded"
+    "|test_scan_sum)"
+)
+
+# The following tests are about deprecated operators.
+backend_test.exclude("(test_scatter_with_axis|test_scatter_without)")
+
+
+# The following tests are too slow with the reference implementation (Conv).
```

Comment:
Is the recommendation now to exclude these models from routine backend testing? They do seem to be very slow:
```
9335.38s call     test_backend.py::OnnxBackendRealModelTest::test_vgg19_cuda
8523.93s call     test_backend.py::OnnxBackendRealModelTest::test_resnet50_cuda
6091.74s call     test_backend.py::OnnxBackendRealModelTest::test_densenet121_cuda
2532.54s call     test_backend.py::OnnxBackendRealModelTest::test_inception_v2_cuda
1631.90s call     test_backend.py::OnnxBackendRealModelTest::test_inception_v1_cuda
595.61s call     test_backend.py::OnnxBackendRealModelTest::test_squeezenet_cuda
530.00s call     test_backend.py::OnnxBackendRealModelTest::test_zfnet512_cuda
428.94s call     test_backend.py::OnnxBackendRealModelTest::test_shufflenet_cuda
200.11s call     test_backend.py::OnnxBackendRealModelTest::test_bvlc_alexnet_cuda
```

Reply:
I made a PR to fix that issue: https://github.com/onnx/onnx/pull/4960. Not merged yet. They are slow because the reference implementation of operator Conv is slow.